pennylane/vqe/vqe.py

```diff
@@ -129,6 +160,119 @@ def __str__(self):
 
         return "\n+ ".join(terms)
 
+    def obs_data(self):
+        r"""Extracts the data from each Tensor/Observable in a qml.Hamiltonian object.
+
+        Returns:
+            (set): Set of coefficients, paired with sets of the observables tensored to make each term.
+        """
+        data = set()
+
+        for co, op in zip(*self.terms):
+            obs = op.non_identity_obs if isinstance(op, Tensor) else [op]
+            tensor = []
+            for ob in obs:
+                parameters = tuple(
+                    param.tostring() for param in ob.parameters
+                )  # Converts params into hashable type
+                tensor.append((ob.name, ob.wires, parameters))
+            data.add((co, frozenset(tensor)))
+
+        return data
+
+    def compare(self, H):
+        r"""Compares two qml.Hamiltonian objects/Observables/Tensors to determine if they are equivalent.
+
+        Returns:
+            (bool): True if the coefficients and operations of the Hamiltonians are equivalent, False otherwise.
+        """
+        val = False
+        if isinstance(H, Hamiltonian):
+            val = self.obs_data() == H.obs_data()
+        if isinstance(H, (Tensor, Observable)):
+            val = self.obs_data() == {(1, frozenset(H.obs_data()))}
+
+        return val
+
+    def __matmul__(self, H):
+        r"""The tensor product operation between qml.Hamiltonian objects/Tensors/Observables.
+        """
+        coeffs1 = self.coeffs.copy()
+        terms1 = self.ops.copy()
+
+        coeffs = []
+        terms = []
+
+        if isinstance(H, Hamiltonian):
+            coeffs2 = H.coeffs
+            terms2 = H.ops
+
+            coeffs = [c[0] * c[1] for c in itertools.product(coeffs1, coeffs2)]
+            term_list = itertools.product(terms1, terms2)
+            terms = [qml.operation.Tensor(t[0], t[1]) for t in term_list]
+
+        if isinstance(H, (Tensor, Observable)):
+            coeffs = coeffs1
+            terms = [term @ H for term in terms1]
+
+        return qml.Hamiltonian(coeffs, terms, simplify=True)
```

Comment:
Should raise an exception here, in case `H` is neither a tensor/observable/Hamiltonian!